### PR TITLE
Added replica election to Ruler HA proposal

### PR DESCRIPTION
**Pull request by soonping-amzn**
**What this PR does**:  
Adds replica election to the Ruler HA proposal to address cost, monitoring and operational problems arising from duplicated evaluations. Removed ALERTS_FOR_STATE section due to it no longer being needed with elected replicas.

**Which issue(s) this PR fixes**:
This change does not fix any issues.

**Checklist**
- [ ] Tests updated
- [X] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: docs/proposals/ruler-ha.md [hunk 1]
```diff
@@ -18,51 +18,142 @@ The original issue is [#4435](https://github.com/cortexproject/cortex/issues/443
...
 
 [PR #4712](https://github.com/cortexproject/cortex/pull/4712) [open]
 
-### Adjustments for ALERTS_FOR_STATE
+### Elected ruler replicas for single iteration evaluation
 
-When an alert fires in a rulegroup shared across multiple rulers, the alert on each ruler will have slightly different timestamps due to the differences in evaluation time on each ruler.  This results in the ALERTS_FOR_STATE metric being deduplicated with the "duplicate sample for timestamp" error.  To prevent this, we synchronize Alert.activeAt for the same alerts in different rulers using a post-processing hook.
+With replication enabled, multiple rulers will evaluate the same rule group concurrently.  This will result in duplicated work, leading to duplication of the following:
```
Comment (alvinlin123) [inline docs/proposals/ruler-ha.md:47]: I think the proposal would be easier to follow if line 47 to 55 is moved to another section called "why not evaluate rules concurrently in multiple rulers"

File: docs/proposals/ruler-ha.md [hunk 2]
```diff
@@ -18,51 +18,142 @@ The original issue is [#4435](https://github.com/cortexproject/cortex/issues/443
...
 
-A replication factor greater than 1 will result in the same rules being evaluated by multiple ruler instances.
+A replication factor greater than 1 will result in the same rules being available for evaluation by multiple ruler instances.
 
-This redundancy will allow for missed or skipped rule evaluations from single ruler outages to be covered by other instances evaluating the same rules.
+This redundancy will allow for missed or skipped rule evaluations from single ruler outages to be covered by other instances able to evaluate the same rule groups.
 
 There is also the question of duplicate metrics generated by replicated evaluation.  We do not expect this to be a problem, as all replicas must use the same [slotted intervals](https://github.com/prometheus/prometheus/blob/b878527151e6503d24ac5b667b86e8794eb79ff7/rules/manager.go#L509) when evaluating rules, which should result in the same timestamp applying to metrics generated by each replica, with the samples being deduplicated in the ingestion path.
```
Comment (alvinlin123) [inline docs/proposals/ruler-ha.md:41]: Is this still applicable with the new approach?

**Review by friedrichg — COMMENTED** [review]
https://github.com/cortexproject/cortex/blob/58b59f471a6f4d4d19315aee41476e64326bf44e/pkg/util/validation/limits.go#L41
we need ruler_accept_ha_samples as an override
right?

File: docs/proposals/ruler-ha.md [hunk 3]
```diff
@@ -18,51 +18,141 @@ The original issue is [#4435](https://github.com/cortexproject/cortex/issues/443
...
+|--------------|--------------|
+| ${userID}/${namespace}/${groupName} | ReplicaDesc with Replica == ruler pod ID |
+
+#### Configuration changes
+
+Since we will be essentially copying distributor's HA tracker implementation, we will add the same configuration parameters to ruler:
+
+```
+# Ruler HA tracking is automatically enabled when replication_factor is > 1
+ha_tracker:
```
[the remark this answers is not in the document]
Comment (soonping-amzn) [reply docs/proposals/ruler-ha.md:73]: ###  ruler_accept_ha_samples

I don't think we need this.  This parameter's purpose in distributor is related to processing of cluster/replica labels in incoming samples. However, in the ruler context, cluster/replica information is derived from the pod ID and rule group identifiers, which are part of the execution context[^1] during rule group evaluation.

###  enable_ha_tracker

I had originally intended to leave this parameter out, and rely solely on replication_factor as the parameter that enables/disables ruler's HA tracker, but later changed my mind once I realized that explicitly allowing users to enable HA tracking allows for safer migrations from replication_factor=1 to replication_factor>1 using a 2-phase deployment:

1. deploy with replication_factor=1, ruler.enable_ha_tracker=true
1. deploy with replication_factor=3

Hence, I'm adding this parameter back in.  Thanks for pointing this out!

[^1]: I'm referring to execution contexts in the general sense, not GoLang Contexts.

**Pull request closed without merging**